test/alternator/conftest.py

```diff
@@ -59,6 +59,46 @@ def pytest_collection_modifyitems(config, items):
         if "veryslow" in item.keywords:
             item.add_marker(skip_veryslow)
 
+# When testing Alternator running with --alternator-enforce-authorization=1,
+# we need to find a valid username and secret key to use in the connection.
+# Alternator allows any CQL role as the username any CQL role, and the key
+# is that role's password's salted hash. We can read a valid role/hash
+# from the appropriate system table, but can't do it with Alternator (because
+# we don't know yet the secret key!), so we need to do it with CQL.
+get_valid_alternator_role_cache = {}
+def get_valid_alternator_role(url):
+    if url in get_valid_alternator_role_cache:
+        return get_valid_alternator_role_cache[url]
+    import cassandra.cluster
+    import cassandra.auth
+    import urllib.parse
+    host = urllib.parse.urlparse(url).hostname
+    auth_provider = cassandra.auth.PlainTextAuthProvider(
+        username='cassandra', password='cassandra')
+    cluster = cassandra.cluster.Cluster([host],
+        auth_provider=auth_provider,
+        connect_timeout = 60, control_connection_timeout = 60)
+    session = cluster.connect()
```

Comment:
I would recommend using this as a context manager, so it could clear the connection regardless of failure or others

Reply:
I don't want to use more functions (for the reason I explained above), but I agree maybe an except/finally would have made sense. Note that in any case, this function is called only (or almost only, we actually call it directly in one test) in fixture initialization, so if there's a failure, we'll see a bazillion tests fail and whether or not we closed the connection is the least of our worries.

Reply:
it's not new function, it how it can work out of the box:

```python
with cluster.connect() as session:
     # use the session

# out of the context, session should be cleared out (and shutdown called)
```

Reply:
Wait, what? I didn't think that cluster.connect() itself is a context manager (i.e., returns an object with an __exit__). Is it? Is it documented anywhere?

Reply:
I see you're right, /usr/local/lib64/python3.12/site-packages/cassandra/cluster.py in class Cluster, there is indeed a
```
   def __exit__(self, *args):
        self.shutdown()
```
I suddenly also realized that this **is** documented:
> ``Cluster`` and ``Session`` also provide context management functions which implicitly handle shutdown when leaving scope.

I'll use it. Thanks.

Reply:
Oh, but there is a complication :-(
I need to have the "with" on the **cluster**, not on the session (the result of cluster.connect()). According to the documentation, the important part is to shutdown() the cluster, NOT the session.
I guess I can do it both both just to be on the safe side (and have really ugly indented code :-( ).

Reply:
in dtest we do have a wrapper context manager that closes both at the same time.

I think using the Cluster as context manager would suffice in you case (it would also close the sessions)

Reply:
I ended up doing the following (only works since Python 3.10):

```python
    with (
        Cluster([urlparse(url).hostname], auth_provider=auth_provider,
            connect_timeout = 60, control_connection_timeout = 60) as cluster,
        cluster.connect() as session
    ):
```

Reply:
> I think using the Cluster as context manager would suffice in you case (it would also close the sessions)

I agree, and this is what I did in the old code, but since it's not hard to do have them both with-ed (see snippet above), I decided to do both. Who knows, maybe in the future this will become important for some reason...

In any case, it was important to realize that the `with cluster.connect() as session:` that you suggested originally was actually the wrong thing to do, as it was most important to shutdown the cluster - not the session.